Wrapped bindings

mikea opened the pull request:
Wrapped bindings provide higher-level API access to lower-level inner bindings.

To define the binding a wrapper module and inner bindings need to be specified:

```
  bindings = [
    (
      # This binding will be accessible as `env.hn` the usual way.
      name = "hn",
      wrapped = (
        # `wrapBindings` function from hnApi module will be called to create the final binding.
        wrapWith = "hnApi",
        # Inner bindings are passed as `env` to the `wrapBindings` function.
        # In this cases it will be accessible as `env.internet`
        innerBindings = [ (name = "internet", service = "internet") ],
      )
    )
  ],
```

Wrapper module needs to export `wrapBindings` that will be executed during binding creation phase:

```js
export function wrapBindings(env) {
  return {
    async query(q) {
      const response = await env.internet.fetch(/* build request */);
      return await response.json();
    }
  }
}
```

mikea commented:
This is a cleaned up version of #404. I broke down the code in multiple changes and added plenty of documentation. PTAL.

jasnell commented on samples/wrappedBinding/config.capnp:
```diff
@@ -0,0 +1,30 @@
+# This is a demo for wrapped bindings.
```
Nit: would be good to explain in the comment what a "wrapped binding" is

jasnell commented on src/workerd/server/workerd-api.h:
```diff
@@ -116,9 +116,20 @@ class WorkerdApiIsolate final: public Worker::ApiIsolate {
         return *this;
       }
     };
+    struct Wrapped {
```
Nit: A code comment here explaining what this is would be helpful.

jasnell reviewed (approved):
Initial read through LGTM but let's have @kentonv review and sign off also.

mikea commented on src/workerd/server/workerd.capnp:
```diff
@@ -420,6 +423,24 @@ struct Worker {
+
+      wrapWith @0 :Text;
+      # Wrapper module nume.
+      # Can be both a module from current worker or a built-in module.
```
I was going to change the code to allow only builtins, but realized there's no easy way to provide builtins to a workerd besides making them real and link with workerd.

I've opened #463 to allow specifying builtins through config.

I'll update this PR once that is in.

jasnell commented on samples/wrappedBinding/hnApi.js:
```diff
@@ -0,0 +1,15 @@
+export function wrapBindings(env) {
+  return {
+    async query(q) {
+      let url = "http://hn.algolia.com/api/v1/search_by_date?";
```
Good idea to add a comment here reminding folks that this points to an external server whose content is out of our control.

mikea commented:
Rebased on top of extensions work and introduced following changes:
- only internal modules can produce wrapped bindings now
- added tests
- made `wrapWithFn` configurable with default value
- reworked the example not to depend on external web sites


I think this addresses all the comments. PTAL.

kentonv commented on src/workerd/server/workerd.capnp:
```diff
@@ -424,6 +427,26 @@ struct Worker {
+
+      moduleName @0 :Text;
+      # Wrapper module nume.
+      # Can be both a module from current worker or a built-in module.
```
Comment needs updating?

kentonv commented on src/workerd/server/workerd.capnp:
```diff
@@ -420,6 +423,24 @@ struct Worker {
+      # Wrapper module nume.
+      # Can be both a module from current worker or a built-in module.
+      # The module will be instantitated during binding initialization phase.
+      # Module needs to exports "wrapBindings(env)" function.
```
This comment was marked resolved but I don't think it was resolved.

mikea replied:
there's a `wrapWithFn` property define to customize the function. Or do you mean that we should use `default` export by default?

kentonv commented:
Ah sorry, I missed `wrapWithFn`. But I dunno, it feels awkward to me, but I can't put my finger on why, it might just be the naming.

Can at least do this:
* Rename `wrapWithFn` -> `entrypoint`. This feels more obvious to me since it uses the terminology we use elsewhere for specifying a module export. (I'd also be OK with `exportName` here, if people don't like `entrypoint`.)
* One of the following:
    * Rename the default `wrapBindings` -> `makeWrappedBinding`, OR
    * Use the default export by default, OR
    * For extreme consistency with Workers, we could make you do something like `export default { makeWrappedBinding(env) { ... } }`, so the entrypoint is an object rather than a function. The potential advantage of this is that we can detect programmatically when an export is intended to be a binding wrapper, just like how today we detect whether an entrypoint is a fetch handler vs. a cron trigger, etc. However, I admittedly don't know that this is actually useful in practice, since obviously extension code and regular worker code don't belong together. So maybe this is a silly idea.

mikea commented:
> Ah sorry, I missed `wrapWithFn`. But I dunno, it feels awkward to me, but I can't put my finger on why, it might just be the naming.
> 
> Can at least do this:
> 
> * Rename `wrapWithFn` -> `entrypoint`. This feels more obvious to me since it uses the terminology we use elsewhere for specifying a module export. (I'd also be OK with `exportName` here, if people don't like `entrypoint`.)
> * One of the following:
>   
>   * Rename the default `wrapBindings` -> `makeWrappedBinding` (so that the function name describes the output, rather than the inputs), OR
>   * Use the default export by default, so there's no default name, OR
>   * For extreme consistency with Workers, we could make you do something like `export default { makeWrappedBinding(env) { ... } }`, so the entrypoint is an object rather than a function. The potential advantage of this is that we can detect programmatically when an export is intended to be a binding wrapper, just like how today we detect whether an entrypoint is a fetch handler vs. a cron trigger, etc. However, I admittedly don't know that this is actually useful in practice, since obviously extension code and regular worker code don't belong together. So maybe this is a silly idea.

Done. Renamed `wrapWithFn` to `entrypoint`. And since default export is represented as `default` property in module namespace, default export works with `default` default value :) We already use this fact in `worker.c++`.

kentonv commented:
Don't forget to squash.
